File: async-logger/src/agent.ts
import "dotenv/config";
import { getEnvOrThrow } from "./utils.js";
import { Payments, type EnvironmentName } from "@nevermined-io/payments";
import OpenAI from "openai";

const OPENAI_API_KEY = getEnvOrThrow("OPENAI_API_KEY");
const NVM_API_KEY = getEnvOrThrow("NVM_API_KEY");
const NVM_AGENT_ID = getEnvOrThrow("NVM_AGENT_ID");
const NVM_PLAN_ID = getEnvOrThrow("NVM_PLAN_ID");
const NVM_ENVIRONMENT = getEnvOrThrow("NVM_ENVIRONMENT");

// This is not meant to be a full fledged example, it is just to show how to use the async logger
// For a complete example of how to monetize an agent with Nevermined, see the financial-agent example
const EXAMPLE_REQUEST_URL = "https://example.com/ask";

const payments = Payments.getInstance({
  nvmApiKey: NVM_API_KEY,
  environment: NVM_ENVIRONMENT as EnvironmentName,
});

const getAccessToken = async () => {
  const balanceInfo = await payments.plans.getPlanBalance(NVM_PLAN_ID);
  if (!balanceInfo.isSubscriber || balanceInfo.balance <= 0) {
    console.log(
      "User is not a subscriber or has no credits. Purchasing plan..."
    );
    await payments.plans.orderPlan(NVM_PLAN_ID);
  }
  const { accessToken } = await payments.agents.getAgentAccessToken(
    NVM_PLAN_ID,
    NVM_AGENT_ID
  );
  if (!accessToken) {
    throw new Error("Failed to obtain access token");
  }
  return accessToken;
};

const agent = async (accessToken: string) => {
  // validate request is authorized
  console.log("Validating request is authorized...");
  const agentRequest = await payments.requests.startProcessingRequest(
    NVM_AGENT_ID,
    accessToken,
    EXAMPLE_REQUEST_URL,
    "POST"
  );
  if (
    !agentRequest.balance.isSubscriber ||
    agentRequest.balance.balance <= 0n
  ) {
    throw new Error("Payment Required");
  }

  // Initialize the async logger
  const logger = payments.observability.withAsyncLogger(
    {
      openAI: OpenAI,
    },
    agentRequest
  );
  logger.init();

  console.log("Making OpenAI LLM call...");
  const openai = new OpenAI({
    apiKey: OPENAI_API_KEY,
  });
  const completion = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [{ role: "user", content: "Hello, world!" }],
  });

  // redeem credits
  console.log("Redeeming credits...");
  await payments.requests.redeemCreditsFromRequest(
    agentRequest.agentRequestId,
    accessToken,
    1n
  );

  return completion.choices[0]?.message?.content;
};

const main = async () => {
  console.log("Getting access token...");
  const accessToken = await getAccessToken();
  console.log("Calling agent...");
  const response = await agent(accessToken);
  console.log(response);
};

main().then(() => {
  process.exit(0);
});

File: http-simple-agent-py/src/agent.py
"""
Protected HTTP Agent - FastAPI server with Nevermined payment middleware.

This demonstrates how to protect an endpoint using the x402 protocol.
The /ask endpoint requires a valid x402 access token and burns credits.

x402 HTTP Transport Headers:
- Client sends token in: `payment-signature` header
- Server returns 402 with: `payment-required` header (base64-encoded)
"""

import os
import sys

# IMPORTANT: Load environment variables BEFORE importing payments_py
# so that env vars are available during SDK initialization
from dotenv import load_dotenv

load_dotenv()

import uvicorn
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from openai import OpenAI
from pydantic import BaseModel

from payments_py import Payments, PaymentOptions
from payments_py.x402.fastapi import PaymentMiddleware, X402_HEADERS

# Configuration
PORT = int(os.getenv("PORT", "3000"))
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY", "")
NVM_API_KEY = os.getenv("NVM_API_KEY", "")
NVM_ENVIRONMENT = os.getenv("NVM_ENVIRONMENT", "sandbox")
NVM_PLAN_ID = os.getenv("NVM_PLAN_ID", "")

# Validate required environment variables
if not OPENAI_API_KEY:
    print("OPENAI_API_KEY is required. Set it in .env file.")
    sys.exit(1)

if not NVM_API_KEY or not NVM_PLAN_ID:
    print("NVM_API_KEY and NVM_PLAN_ID are required for payment protection.")
    sys.exit(1)

# Initialize OpenAI client
openai_client = OpenAI(api_key=OPENAI_API_KEY)

# Initialize Nevermined Payments SDK
payments = Payments.get_instance(
    PaymentOptions(nvm_api_key=NVM_API_KEY, environment=NVM_ENVIRONMENT)
)

# Create FastAPI app
app = FastAPI(
    title="HTTP Simple Agent (Python)",
    description="FastAPI server with x402 payment middleware",
)


# Request model
class AskRequest(BaseModel):
    query: str


# Add payment middleware
app.add_middleware(
    PaymentMiddleware,
    payments=payments,
    routes={
        "POST /ask": {
            "plan_id": NVM_PLAN_ID,
            "credits": 1,
        },
    },
)


@app.post("/ask")
async def ask(request: Request, body: AskRequest) -> JSONResponse:
    """Send a query to the AI (protected by payment middleware)."""
    try:
        if not body.query or not isinstance(body.query, str):
            return JSONResponse(
                status_code=400,
                content={"error": "Missing or invalid 'query' field"},
            )

        completion = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {
                    "role": "system",
                    "content": "You are a helpful assistant. Be concise and informative.",
                },
                {"role": "user", "content": body.query},
            ],
            max_tokens=500,
        )

        response_text = (
            completion.choices[0].message.content
            if completion.choices
            else "No response generated"
        )

        return JSONResponse(content={"response": response_text})

    except Exception as error:
        print(f"Error in /ask: {error}")
        return JSONResponse(
            status_code=500,
            content={"error": "Internal server error"},
        )


@app.get("/health")
async def health() -> JSONResponse:
    """Health check endpoint (unprotected)."""
    return JSONResponse(content={"status": "ok"})


def main():
    """Run the FastAPI server."""
    print(f"Protected HTTP Agent running on http://localhost:{PORT}")
    print(f"\nPayment protection enabled for POST /ask")
    print(f"Plan ID: {NVM_PLAN_ID}")
    print(
        f"\nTo test, send requests with x402 token in '{X402_HEADERS['PAYMENT_SIGNATURE']}' header."
    )

    uvicorn.run(app, host="0.0.0.0", port=PORT)


if __name__ == "__main__":
    main()
